mpi_overlap/jacobi.cpp

```diff
@@ -148,19 +150,25 @@ int main(int argc, char* argv[]) {
     const bool csv = get_arg(argv, argv + argc, "-csv");
     const bool use_hp_streams = get_arg(argv, argv + argc, "-use_hp_streams");
 
-    int local_rank = -1;
-    {
+    if(num_devices > 1) {
+        int local_rank = -1;
+     
         MPI_Comm local_comm;
         MPI_CALL(MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, rank, MPI_INFO_NULL,
                                      &local_comm));
 
         MPI_CALL(MPI_Comm_rank(local_comm, &local_rank));
 
         MPI_CALL(MPI_Comm_free(&local_comm));
+
+        CUDA_RT_CALL(cudaSetDevice(local_rank));
+        CUDA_RT_CALL(cudaFree(0));
     }
 
-    CUDA_RT_CALL(cudaSetDevice(local_rank));
-    CUDA_RT_CALL(cudaFree(0));
+    else {
+        CUDA_RT_CALL(cudaSetDevice(0));
+        CUDA_RT_CALL(cudaFree(0));
+    }
```

Comment:
Please use `CUDA_RT_CALL(cudaSetDevice(local_rank%num_devices));` it avoids the if/else branch minimizing the new code added but more importantly it also covers the case if a GPU is oversubscribed with MPS.

This applies to MPI and MPI overlap not to the NCCL and NVSHMEM variants.